src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -398,9 +397,312 @@ pub trait TimestampProvider {
+        // Generally, we want to choose the freshest timestamp possible, although there are exceptions
+        // when we either want a maximally *stale* timestamp, or we want to protect other queries from
+        // a recklessly advanced timestamp.
+        let preferences = {
+            // Counter-intuitively, the only `when` that allows `can_advance_to_upper` is `Immediately`,
+            // and not `FreshestTableWrite`. This is because `FreshestTableWrite` instead imposes a lower
+            // bound through the `oracle_read_ts`, and then requires the stalest valid timestamp.
+
+            if when.can_advance_to_upper()
+                && (isolation_level == &IsolationLevel::Serializable || timeline.is_none())
+            {
+                Preference::FreshestAvailable
+            } else {
+                Preference::StalestValid
+            }
+
+            // TODO: `StrongSessionSerializable` has a different set of preferences that starts to tease
+            // out the trade-off between freshness and responsiveness. I think we don't yet know enough
+            // to properly frame these preferences, though they are clearly aimed at the right concerns.
+        };
+
+        // Determine a candidate based on constraints and preferences.
+        let constraint_candidate = {
+            let mut candidate = Timestamp::minimum();
+            candidate.advance_by(constraints.lower_bound().borrow());
+            // If we have a preference to be the freshest available, advance to the minimum
+            // of the upper bound constraints and the `largest_not_in_advance_of_upper`.
+            if let Preference::FreshestAvailable = preferences {
+                let mut upper_bound = constraints.upper_bound();
+                upper_bound.insert(largest_not_in_advance_of_upper);
+                candidate.advance_by(upper_bound.borrow());
+            }
+            // If the candidate strictly exceeds the upper bound, we didn't have a viable timestamp.
+            if constraints.upper_bound().less_than(&candidate) {
+                coord_bail!(generate_timestamp_not_valid_error_msg(
+                    id_bundle,
+                    compute_instance,
+                    &read_holds,
+                    candidate
+                ));
+            } else {
+                candidate
+            }
+        };
+
+        Ok(RawTimestampSelection {
+            timestamp: constraint_candidate,
+            constraints: Some(constraints),
+            session_oracle_read_ts,
+        })
+    }
+
+    /// Determines the timestamp for a query.
+    ///
+    /// Timestamp determination may fail due to the restricted validity of
+    /// traces. Each has a `since` and `upper` frontier, and are only valid
+    /// after `since` and sure to be available not after `upper`.
+    ///
+    /// The timeline that `id_bundle` belongs to is also returned, if one exists.
+    fn determine_timestamp_for(
```

Comment:
This is perhaps future work, but the way that `determine_timestamp` calls in to this method might be appropriate to blend in with this logic (iirc, it calls this method up to twice, with different isolation levels, to judge what would happen if you relaxed the isolation level). That logic was incorrect when last I looked at it (it had an `await` between the calls, so the two determinations might be unrelated). But, flagging for near future thoughts.